IoT Plug and Play GA support

daisukeiot opened the pull request:
- Add a device model and code for the model
- Re-factored LED code
- Refactored various code to keep the binary size < 0xAF00 bytes
- Set indent for debug print
- Remove unnecessary traces
- Consistent coding style (Tab vs space etc)
- Add CLI to configure DPS ID Scope
- Add CLI to control LED over UART
- Add button press as event telemetry

## Purpose
Certify the device for IoT Plug and Play and publish to the device catalog.

## Does this introduce a breaking change?
[ ] Yes
[x] No

## Pull Request Type
What kind of change does this Pull Request introduce?

[x ] Bugfix
[x] Feature
[x ] Code style update (formatting, local variables)
[x ] Refactoring (no functional changes, no api changes)
[ ] Documentation content changes
[ ] Other... Please describe:
```

## How to Test
Flash the FW and run the IoT Plug and Play certification test.

* Test the code
- Use CLI to control LED to make sure properties are reported to IoT Hub
- Provision device using DPS
- Send Commands from Cloud
- Update Writable property to control LED and telemetry interval

## Other Information

daisukeiot committed:
IoT Plug and Play Device Certification (#1)

- Add a device model and code for the model
- Re-factored LED code
- Set indent for debug_printXXXX()
- Remove unnecessary traces traces

ericwolz commented on AzureIotPnpDps.X/Model/pic_iot_wm-1.json:
```diff
@@ -0,0 +1,342 @@
+{
+  "@context": "dtmi:dtdl:context;2",
+  "@id": "dtmi:com:Microchip:PIC_IoT_WM;1",
```
PIC24_IoT_WM ?

daisukeiot replied:
The ID was determined by Randy (and he already published model to public repo) so probably best to stick with this one.

daisukeiot replied:
Per Randy, let's keep ti PIC_IoT_WM

ericwolz replied:
ok

ericwolz commented on AzureIotPnpDps.X/Model/pic_iot_wm-1.json:
```diff
@@ -0,0 +1,342 @@
+  "@context": "dtmi:dtdl:context;2",
+  "@id": "dtmi:com:Microchip:PIC_IoT_WM;1",
+  "@type": "Interface",
+  "displayName": "PIC-IoT WM",
```
```suggestion
  "displayName": "PIC24-IoT WM",
```

daisukeiot replied:
Checking with Randy

daisukeiot replied:
Per Randy, use PIC-IoT without 24

ericwolz commented on AzureIotPnpDps.X/Model/pic_iot_wm-1.json:
```diff
@@ -0,0 +1,342 @@
+  "@id": "dtmi:com:Microchip:PIC_IoT_WM;1",
+  "@type": "Interface",
+  "displayName": "PIC-IoT WM",
+  "description": "Reports temperature, light intensity, and states of the 2 buttons & 4 LEDs.  Provides ability to turn on/off any of the 4 LEDs.",
```
```suggestion
  "description": "Reports device temperature, light intensity, and the current state of the 2 buttons & 4 LEDs.  Provides ability to turn on/off any of the 4 LEDs.",
```

ericwolz commented on AzureIotPnpDps.X/Model/pic_iot_wm-1.json:
```diff
@@ -0,0 +1,342 @@
+      "displayName": {
+        "en": "Blue LED state"
+      },
+      "description": "Returns current state of Blue LED. If True, Blue LED is on and WiFi AP is connected.",
```
```suggestion
      "description": "Returns current state of the Blue LED. If True, the Blue LED is on and the WiFi AP is connected.",
```

ericwolz commented on AzureIotPnpDps.X/Model/pic_iot_wm-1.json:
```diff
@@ -0,0 +1,342 @@
+            },
+            "enumValue": 1,
+            "description": "Blue LED is turned on.",
+            "comment": "Blue LED is On state."
```
```suggestion
            "comment": "Blue LED is in the On state."
```

ericwolz commented on AzureIotPnpDps.X/Model/pic_iot_wm-1.json:
```diff
@@ -0,0 +1,342 @@
+            },
+            "enumValue": 2,
+            "description": "Blue LED is turned Off.",
+            "comment": "Blue LED is Off state."
```
```suggestion
            "comment": "Blue LED is in the Off state."
```

ericwolz commented on AzureIotPnpDps.X/Model/pic_iot_wm-1.json:
```diff
@@ -0,0 +1,342 @@
+      "displayName": {
+        "en": "Green LED state"
+      },
+      "description": "Returns current state of Green LED. If True, Green LED is on and connected to cloud.",
```
```suggestion
      "description": "Returns the current state of the Green LED. If True, Green LED is on and connected to cloud.",
```

ericwolz commented on AzureIotPnpDps.X/Model/pic_iot_wm-1.json:
```diff
@@ -0,0 +1,342 @@
+            },
+            "enumValue": 1,
+            "description": "Green LED is turned on.",
+            "comment": "Green LED is On state."
```
```suggestion
            "comment": "Green LED is in the On state."
```

ericwolz commented on AzureIotPnpDps.X/Model/pic_iot_wm-1.json:
```diff
@@ -0,0 +1,342 @@
+            },
+            "enumValue": 2,
+            "description": "Green LED is turned Off.",
+            "comment": "Green LED is Off state."
```
```suggestion
            "comment": "Green LED is in the Off state."
```

ericwolz commented on AzureIotPnpDps.X/Model/pic_iot_wm-1.json:
```diff
@@ -0,0 +1,342 @@
+            },
+            "enumValue": 1,
+            "description": "Yellow LED is turned on.",
+            "comment": "Yellow LED is On state."
```
```suggestion
            "comment": "Yellow LED is in the On state."
```

ericwolz commented on AzureIotPnpDps.X/Model/pic_iot_wm-1.json:
```diff
@@ -0,0 +1,342 @@
+            },
+            "enumValue": 2,
+            "description": "Yellow LED is turned Off.",
+            "comment": "Yellow LED is Off state."
```
```suggestion
            "comment": "Yellow LED is in the Off state."
```

ericwolz commented on AzureIotPnpDps.X/Model/pic_iot_wm-1.json:
```diff
@@ -0,0 +1,342 @@
+      "displayName": {
+        "en": "Red LED state"
+      },
+      "description": "Current state of Red LED, with string enum data type",
```
```suggestion
      "description": "Current state of the Red LED",
```

ericwolz commented on AzureIotPnpDps.X/Model/pic_iot_wm-1.json:
```diff
@@ -0,0 +1,342 @@
+            },
+            "enumValue": "On",
+            "description": "Red LED is turned on.",
+            "comment": "Red LED is On state."
```
```suggestion
            "comment": "Red LED is in the On state."
```

ericwolz commented on AzureIotPnpDps.X/Model/pic_iot_wm-1.json:
```diff
@@ -0,0 +1,342 @@
+            },
+            "enumValue": "Off",
+            "description": "Red LED is turned Off.",
+            "comment": "Red LED is Off state."
```
```suggestion
            "comment": "Red LED is in the Off state."
```

ericwolz commented on AzureIotPnpDps.X/Model/pic_iot_wm-1.json:
```diff
@@ -0,0 +1,342 @@
+      "displayName": {
+        "en": "Set Telemetry Interval"
+      },
+      "description": "Sets interval to send telemetry in second",
```
```suggestion
      "description": "Sets interval to send telemetry in seconds",
```

ericwolz commented on AzureIotPnpDps.X/Model/pic_iot_wm-1.json:
```diff
@@ -0,0 +1,342 @@
+      "displayName": {
+        "en": "Reboot"
+      },
+      "description": "Reboot PiC IoT with specified delay.  e.g. PT5S for 5 seconds.",
```
```suggestion
      "description": "Reboot PiC24 IoT with the specified delay.  e.g. PT5S for 5 seconds.",
```

ericwolz commented on AzureIotPnpDps.X/Model/pic_iot_wm-1.json:
```diff
@@ -0,0 +1,342 @@
+      "@type": "Command",
+      "name": "getMaxMinReport",
+      "displayName": {
+        "en": "Get Max-Min report"
```
```suggestion
        "en": "Get temperature Max-Min report"
```

ericwolz commented on AzureIotPnpDps.X/Model/pic_iot_wm-1.json:
```diff
@@ -0,0 +1,342 @@
+      "request": {
+        "name": "since",
+        "displayName": "Since",
+        "description": "Period to return the max-min report",
```
```suggestion
        "description": "The start period to return the max-min report from",
```

ericwolz commented on AzureIotPnpDps.X/main.c:
```diff
@@ -7,582 +7,1326 @@
+// used by pin_manager.c
+extern button_press_data_t button_press_data;
+
+extern uint32_t telemetry_interval;
```
```suggestion
extern uint32_t telemetry_interval_seconds;
```

ericwolz commented on AzureIotPnpDps.X/main.c:
```diff
@@ -7,582 +7,1326 @@
+
+extern uint32_t telemetry_interval;
+
+static bool isGetReceived = false;
```
please use snake case to be consistent with other var names.
```suggestion
static bool is_get_received = false;
```

ericwolz commented on AzureIotPnpDps.X/main.c:
```diff
@@ -7,582 +7,1326 @@
+extern uint32_t telemetry_interval;
+
+static bool isGetReceived = false;
+static int rebootDelay = 0;
```
```suggestion
static int reboot_delay_seconds = 0;
```

ericwolz commented on AzureIotPnpDps.X/main.c:
```diff
@@ -7,582 +7,1326 @@
 static int32_t device_temperature_avg_total;
 static uint32_t device_temperature_avg_count = 0;
 static int32_t device_max_temp;
 static int32_t device_min_temp;
```
This needs a default value (MAX INT?). By default it will be zero, so min value will always be at 0.

daisukeiot replied:
Fixed with INT32_MAX and 0.

ericwolz commented on AzureIotPnpDps.X/main.c:
```diff
@@ -7,582 +7,1326 @@
+
+	if (az_span_size(payload) == 0 || (az_span_size(payload) == 2 && az_span_is_content_equal(empty_payload_span, payload)))
+	{
+		debug_printError("  MAIN: Empty Payload");
```
```suggestion
		debug_printError("  MAIN: Empty command Payload");
```

ericwolz commented on AzureIotPnpDps.X/main.c:
```diff
@@ -7,582 +7,1326 @@
+		}
+		else if (jr.token.kind == AZ_JSON_TOKEN_STRING)
+		{
+				RETURN_ERR_IF_FAILED(az_json_token_get_string(&jr.token, reboot_delay, sizeof(reboot_delay), NULL));
```
spacing

daisukeiot replied:
Fixed

ericwolz commented on AzureIotPnpDps.X/main.c:
```diff
@@ -7,582 +7,1326 @@
+
+	memcpy(buffer, &reboot_delay[2], strlen(reboot_delay) - 3);
+	buffer[strlen(reboot_delay) - 3] = '\0';
+	rebootDelay = atoi((const char *)buffer);
```
you can just do

```suggestion
	rebootDelay = atoi((const char *)&reboot_delay[2]);
```
atoi will stop on the first non-numeric value

ericwolz commented on AzureIotPnpDps.X/platform/config/conf_winc.h:
```diff
@@ -22,8 +22,8 @@ extern "C" {
 //#define CFG_MAIN_WLAN_SSID "GUESTNETWORK"
-#define CFG_MAIN_WLAN_SSID ""
-    
+//#define CFG_MAIN_WLAN_SSID ""
```
delete this line

daisukeiot committed:
Update AzureIotPnpDps.X/Model/pic_iot_wm-1.json

Co-authored-by: Eric Wolz <ericwol@microsoft.com>

daisukeiot committed:
Update AzureIotPnpDps.X/Model/pic_iot_wm-1.json

Co-authored-by: Eric Wolz <ericwol@microsoft.com>

daisukeiot committed:
Update AzureIotPnpDps.X/Model/pic_iot_wm-1.json

Co-authored-by: Eric Wolz <ericwol@microsoft.com>

daisukeiot committed:
Update AzureIotPnpDps.X/main.c

Co-authored-by: Eric Wolz <ericwol@microsoft.com>

daisukeiot committed:
Update AzureIotPnpDps.X/main.c

Co-authored-by: Eric Wolz <ericwol@microsoft.com>

daisukeiot committed:
Update AzureIotPnpDps.X/Model/pic_iot_wm-1.json

Co-authored-by: Eric Wolz <ericwol@microsoft.com>

daisukeiot committed:
Update AzureIotPnpDps.X/Model/pic_iot_wm-1.json

Co-authored-by: Eric Wolz <ericwol@microsoft.com>

daisukeiot committed:
Update AzureIotPnpDps.X/Model/pic_iot_wm-1.json

Co-authored-by: Eric Wolz <ericwol@microsoft.com>

daisukeiot committed:
Update AzureIotPnpDps.X/Model/pic_iot_wm-1.json

Co-authored-by: Eric Wolz <ericwol@microsoft.com>

daisukeiot committed:
Update AzureIotPnpDps.X/Model/pic_iot_wm-1.json

Co-authored-by: Eric Wolz <ericwol@microsoft.com>

daisukeiot committed:
Update AzureIotPnpDps.X/Model/pic_iot_wm-1.json

Co-authored-by: Eric Wolz <ericwol@microsoft.com>

daisukeiot committed:
Update AzureIotPnpDps.X/Model/pic_iot_wm-1.json

Co-authored-by: Eric Wolz <ericwol@microsoft.com>

daisukeiot committed:
Update AzureIotPnpDps.X/Model/pic_iot_wm-1.json

Co-authored-by: Eric Wolz <ericwol@microsoft.com>

daisukeiot committed:
Update AzureIotPnpDps.X/Model/pic_iot_wm-1.json

Co-authored-by: Eric Wolz <ericwol@microsoft.com>

daisukeiot committed:
Update AzureIotPnpDps.X/Model/pic_iot_wm-1.json

Co-authored-by: Eric Wolz <ericwol@microsoft.com>

daisukeiot committed:
Update AzureIotPnpDps.X/main.c

Co-authored-by: Eric Wolz <ericwol@microsoft.com>

daisukeiot committed:
Update AzureIotPnpDps.X/main.c

Co-authored-by: Eric Wolz <ericwol@microsoft.com>

daisukeiot committed:
Update main.c

daisukeiot committed:
Update AzureIotPnpDps.X/Model/pic_iot_wm-1.json

Co-authored-by: Eric Wolz <ericwol@microsoft.com>

daisukeiot committed:
Delete unused variable

daisukeiot committed:
Update AzureIotPnpDps.X/Model/pic_iot_wm-1.json

Co-authored-by: Eric Wolz <ericwol@microsoft.com>

daisukeiot committed:
Update pic_iot_wm-1.json

daisukeiot committed:
change variable names

daisukeiot committed:
Fix Min and Max temperature tracking

Clean up traces

danewalton-msft commented on AzureIotPnpDps.X/main.c:
```diff
@@ -7,582 +7,1327 @@
+//
+// * LEDs *
+//        | On                   | Off                      | Fast Blink (100ms)           | Slow Blink (400ms)
+// Blue	  | AP Connected         | AP Disconnected          | WiFi Connectin in progress   | 
```
nit: tab snuck in here right after `Blue`. Change to a space

danewalton-msft replied:
```suggestion
// Blue	  | AP Connected         | AP Disconnected          | WiFi Connection in progress   | 
```

danewalton-msft commented on AzureIotPnpDps.X/main.c:
```diff
@@ -7,582 +7,1327 @@
+// * IoT Plug and Play Values *
+// The model id is the unique identifier for device model (also called the Digital Twins Model Identifier or DTMI)
+// The device Model defines the capability of your device. The functionality of the device should match what
+// is described in the corresponding device model while following IoT Plug and Play convension. 
```
```suggestion
// is described in the corresponding device model while following IoT Plug and Play convention. 
```

danewalton-msft commented on AzureIotPnpDps.X/main.c:
```diff
@@ -7,582 +7,1327 @@
+
+// * IoT Plug and Play Values *
+// The model id is the unique identifier for device model (also called the Digital Twins Model Identifier or DTMI)
+// The device Model defines the capability of your device. The functionality of the device should match what
```
for consistency...doesn't seem to be uppercase elsewhere
```suggestion
// The device model defines the capability of your device. The functionality of the device should match what
```

danewalton-msft commented on AzureIotPnpDps.X/main.c:
```diff
@@ -7,582 +7,1327 @@
 
-// Plug and Play Telemetry Values
+// IoT Plug and Play temperature and light sensor telemetry
+// Frequency of telemetry is determined by telemetryInterval desired property (or writable property)
```
nit:
The language [here](https://github.com/Azure/opendigitaltwins-dtdl/blob/master/DTDL/v2/dtdlv2.md#property) makes little mention of "desired" and instead is writable, read-write, read-only, etc.
```suggestion
// Frequency of telemetry is determined by the writable property telemetryInterval.
```

danewalton-msft commented on AzureIotPnpDps.X/main.c:
```diff
@@ -7,582 +7,1327 @@
+	switch (property_response->response_type)
+	{
+		// A response from a twin GET publish message with the twin document as a payload.
+		case AZ_IOT_HUB_CLIENT_TWIN_RESPONSE_TYPE_GET:
```
```suggestion
		case AZ_IOT_PNP_CLIENT_PROPERTY_RESPONSE_TYPE_GET:
```

danewalton-msft commented on AzureIotPnpDps.X/main.c:
```diff
@@ -7,582 +7,1327 @@
+			break;
+
+		// An update to the desired properties with the properties as a JSON payload.
+		case AZ_IOT_HUB_CLIENT_TWIN_RESPONSE_TYPE_DESIRED_PROPERTIES:
```
```suggestion
		case AZ_IOT_PNP_CLIENT_PROPERTY_RESPONSE_TYPE_DESIRED_PROPERTIES:
```

danewalton-msft commented on AzureIotPnpDps.X/main.c:
```diff
@@ -7,582 +7,1327 @@
+
+		// A response from a twin reported properties publish message. With a successful update of
+		// the reported properties, the payload will be empty and the status will be 204.
+		case AZ_IOT_HUB_CLIENT_TWIN_RESPONSE_TYPE_REPORTED_PROPERTIES:
```
```suggestion
		case AZ_IOT_PNP_CLIENT_PROPERTY_RESPONSE_TYPE_REPORTED_PROPERTIES:
```

danewalton-msft commented on AzureIotPnpDps.X/main.c:
```diff
@@ -7,582 +7,1327 @@
+*	Add a JSON key-value pair with int32 data
+*	e.g. "property_name" : property_val (number)
+**********************************************/
+static az_result append_jason_property_int32(
```
here and elsewhere. should be able to do a find all and replace
```suggestion
static az_result append_json_property_int32(
```

danewalton-msft commented on AzureIotPnpDps.X/main.c:
```diff
@@ -7,582 +7,1327 @@
+	RETURN_ERR_IF_FAILED(az_json_writer_append_string(jw, ack_description));
+	RETURN_ERR_IF_FAILED(az_json_writer_append_end_object(jw));
+	return AZ_OK;
+}
```
this is now a pnp api. Please see [`az_iot_pnp_client_property_builder_begin_reported_status()`](https://github.com/Azure/azure-sdk-for-c/blob/3d9f1a3d530ef5fa12cf481ce7e6a4b798e286d5/sdk/inc/azure/iot/az_iot_pnp_client.h#L631-L637) and [`az_iot_pnp_client_property_builder_end_reported_status()`](https://github.com/Azure/azure-sdk-for-c/blob/3d9f1a3d530ef5fa12cf481ce7e6a4b798e286d5/sdk/inc/azure/iot/az_iot_pnp_client.h#L654-L656). Also [here](https://github.com/Azure/azure-sdk-for-c/blob/3d9f1a3d530ef5fa12cf481ce7e6a4b798e286d5/sdk/samples/iot/paho_iot_pnp_sample.c#L840-L865) for usage.

daisukeiot replied:
https://github.com/daisukeiot/Microchip-PIC-IoT-Wx-1/commit/674a6d446833826ee90159f8dd9934b16bdd3d68

danewalton-msft commented on AzureIotPnpDps.X/main.c:
```diff
@@ -7,582 +7,1327 @@
+static az_result build_error_response_payload(
+	az_span response,
+	az_span error_string,
+	az_span* response_payload)
```
above function has `_span` as a suffix. consider:
```suggestion
	az_span response_span,
	az_span error_string_span,
	az_span* response_payload_span)
```

daisukeiot replied:
Replaced all variables with _span suffix

[commit](https://github.com/daisukeiot/Microchip-PIC-IoT-Wx-1/commit/64508ae802742fd040c4263817d02b3a719ddeaa)

danewalton-msft commented on AzureIotPnpDps.X/main.c:
```diff
@@ -7,582 +7,1327 @@
+	az_json_reader jr;
+	int32_t incoming_since_value_len;
+
+	//we don't actually respect this sinceStr to keep the sample simple,
```
nit:
```suggestion
	// we don't actually respect this sinceStr to keep the sample simple,
```

danewalton-msft commented on AzureIotPnpDps.X/Model/pic_iot_wm-1.json:
```diff
@@ -0,0 +1,342 @@
+        "name": "delay",
+        "schema": "duration",
+        "displayName": "Reboot Delay",
+        "description": "Number of seconds to delay reboot. e.g. PT5S for 5 seconds."
```
curious why make it this complicated?

daisukeiot replied:
What do you mean?  Having a delay?

daisukeiot replied:
Without delay, device does not have time to respond to the command.  Cloud side will think the command failed (including certification)

danewalton-msft commented on AzureIotPnpDps.X/main.c:
```diff
@@ -7,582 +7,1327 @@
-  RETURN_ERR_WITH_MESSAGE_IF_FAILED(rc, "Could not get the telemetry topic");
+	az_span command_response_span = AZ_SPAN_FROM_BUFFER(commands_response_payload);
+	uint16_t return_code;
+	az_result azRet;
```
nit: throughout the code base, the coding style is snake case
```suggestion
	az_result az_ret;
```

daisukeiot replied:
changed all az_result variables to az_ret

[commit](https://github.com/daisukeiot/Microchip-PIC-IoT-Wx-1/commit/e8683cbe4bce84d1a46556a24856d82750f067fa)

danewalton-msft commented on AzureIotPnpDps.X/main.c:
```diff
@@ -7,582 +7,1327 @@
+  do {                                              \
+    if (az_result_failed(ret)) {                    \
+      debug_printError(msg);                        \
+      return ret;                                   \
```
ahh good catch here thank you

danewalton-msft commented on AzureIotPnpDps.X/main.c:
```diff
@@ -7,582 +7,1327 @@
+				// if response is empty, payload was not in the right format.
+				if (az_result_failed(azRet = build_error_response_payload(command_response_span, span_error_no_payload_reboot, &command_response_span)))
+				{
+					debug_printError("  MAIN: Failed to build error response. AZ Result %d", azRet);
```
we've used `0x%08x` [in the past](https://github.com/Azure/azure-sdk-for-c/blob/3d9f1a3d530ef5fa12cf481ce7e6a4b798e286d5/sdk/samples/iot/iot_sample_common.c#L68) which works well to correlate to an az_result value.
```suggestion
					debug_printError("  MAIN: Failed to build error response. AZ Result 0x%08x", azRet);
```

daisukeiot replied:
changed all instances with 0x%08x

[commit](https://github.com/daisukeiot/Microchip-PIC-IoT-Wx-1/commit/e8683cbe4bce84d1a46556a24856d82750f067fa)

danewalton-msft commented on AzureIotPnpDps.X/main.c:
```diff
@@ -7,582 +7,1327 @@
+				}
+			}
+			RETURN_ERR_IF_FAILED(az_json_reader_next_token(&jr));
+		}
```
this section can use the new pnp API found [here](https://github.com/Azure/azure-sdk-for-c/blob/3d9f1a3d530ef5fa12cf481ce7e6a4b798e286d5/sdk/inc/azure/iot/az_iot_pnp_client.h#L717-L722). Usage is [here](https://github.com/Azure/azure-sdk-for-c/blob/3d9f1a3d530ef5fa12cf481ce7e6a4b798e286d5/sdk/samples/iot/paho_iot_pnp_sample.c#L546-L581)

daisukeiot replied:
Done

[commit](https://github.com/daisukeiot/Microchip-PIC-IoT-Wx-1/commit/361f25c335a4354cb64c90d56d204d08b53f3b16)

danewalton-msft commented on AzureIotPnpDps.X/main.c:
```diff
@@ -7,582 +7,1327 @@
+			RETURN_ERR_IF_FAILED(az_json_reader_skip_children(&jr));
+		}
+		RETURN_ERR_IF_FAILED(az_json_reader_next_token(&jr));
+	}
```
this section as well as said above. can make use of `az_iot_pnp_client_property_get_next_component_property()`

daisukeiot replied:
done

[commit](https://github.com/daisukeiot/Microchip-PIC-IoT-Wx-1/commit/361f25c335a4354cb64c90d56d204d08b53f3b16)

danewalton-msft commented on AzureIotPnpDps.X/main.c:
```diff
@@ -7,582 +7,1327 @@
+	//
+	while (jr.token.kind != AZ_JSON_TOKEN_END_OBJECT)
+	{
+		if (az_json_token_is_text_equal(&jr.token, span_desired_property_version))
```
we have an API which will get the property version called `az_iot_pnp_client_property_get_property_version()`. Usage [here](https://github.com/Azure/azure-sdk-for-c/blob/3d9f1a3d530ef5fa12cf481ce7e6a4b798e286d5/sdk/samples/iot/paho_iot_pnp_sample.c#L531-L540)

daisukeiot replied:
changed 

[commit](https://github.com/daisukeiot/Microchip-PIC-IoT-Wx-1/commit/361f25c335a4354cb64c90d56d204d08b53f3b16)

danewalton-msft commented on AzureIotPnpDps.X/main.c:
```diff
@@ -7,582 +7,1327 @@
+		char request_id_buf[50];
+		az_span_to_str(request_id_buf, sizeof(request_id_buf), property_response.request_id);
+		debug_printInfo("  MAIN: Property request, request_id:%s, status: %d", request_id_buf, property_response.status);
+	}
```
nit: newline between logical blocks. different `if/else-if/else` blocks blend together when not separated.

danewalton-msft commented on AzureIotPnpDps.X/main.c:
```diff
@@ -7,582 +7,1327 @@
+
+	// save flags in case user pressed buttons very fast, and for error case.
+	bool sw0 = button_press_data.flag.sw0 == 1 ? true : false;
+	bool sw1 = button_press_data.flag.sw1 == 1 ? true : false;
```
booleans are often easier to use with verbs attached to the name
```suggestion
	bool sw0_pressed = button_press_data.flag.sw0 == 1 ? true : false;
	bool sw1_pressed = button_press_data.flag.sw1 == 1 ? true : false;
```

leads to below suggestions

daisukeiot replied:
Done
[commit](https://github.com/daisukeiot/Microchip-PIC-IoT-Wx-1/commit/78a33ee2f255993b7b48e1c06e1b697ec03caf30)

danewalton-msft commented on AzureIotPnpDps.X/main.c:
```diff
@@ -7,582 +7,1327 @@
+	// clear the flags
+	button_press_data.flag.AsUSHORT = 0;
+
+	if (sw0 != true && sw1 != true)
```
```suggestion
	if (!sw0_pressed && !sw1_pressed)
```

daisukeiot replied:
Done
[commit](https://github.com/daisukeiot/Microchip-PIC-IoT-Wx-1/commit/78a33ee2f255993b7b48e1c06e1b697ec03caf30)

danewalton-msft commented on AzureIotPnpDps.X/main.c:
```diff
@@ -7,582 +7,1327 @@
+
+	RETURN_IF_FAILED(start_json_object(&jw, button_event_span));
+
+	if (sw0 == true)
```
```suggestion
	if (sw0_pressed)
```

danewalton-msft commented on AzureIotPnpDps.X/main.c:
```diff
@@ -7,582 +7,1327 @@
+		RETURN_IF_FAILED(append_button_press_telemetry(&jw, span_event_name_button_sw0, button_press_data.sw0_press_count));
+	}
+
+	if (sw1 == true)
```
```suggestion
	if (sw1_pressed)
```

danewalton-msft commented on AzureIotPnpDps.X/main.c:
```diff
@@ -7,582 +7,1327 @@
+	twin_properties_t* twin_properties_ptr;
+	twin_properties_t twin_properties_local;
+
+	bool bForceSync = false; // force LED status synchronization if this request is coming from Twin Get
```
nit: consistent code style
```suggestion
	bool b_force_sync = false; // force LED status synchronization if this request is coming from Twin Get
```

daisukeiot replied:
done

[commit](https://github.com/daisukeiot/Microchip-PIC-IoT-Wx-1/commit/4150d1aef0b0106fd1b4799cdb5b2d4520083ebf)

danewalton-msft commented on AzureIotPnpDps.X/main.c:
```diff
@@ -7,582 +7,1327 @@
+
+	debug_printInfo("  MAIN: %s() led_change %x", __func__, led_status.change_flag.AsUSHORT);
+
+  // if this is from Get Twin, update according to Desired Property
```
spaces here instead of tabs

daisukeiot replied:
done

[commit](https://github.com/daisukeiot/Microchip-PIC-IoT-Wx-1/commit/4150d1aef0b0106fd1b4799cdb5b2d4520083ebf)

danewalton-msft commented on AzureIotPnpDps.X/mqtt_packetPopulation/mqtt_iotprovisioning_packetPopulate.c:
```diff
@@ -28,13 +28,14 @@ pf_MQTT_CLIENT pf_mqtt_iotprovisioning_client = {
 extern const az_span device_model_id;
 extern uint8_t device_id_buf[100];
 extern az_span device_id;
+uint8_t atca_dps_id_scope[12];
```
what is `atca`?

daisukeiot replied:
I don't know what it stands for but it's Microchip's crypto library.

https://microchiptech.github.io/cryptoauthlib/a00188.html#a22bd6643f31f1d75dc3e7ea939f468cd

danewalton-msft commented on AzureIotPnpDps.X/mqtt_packetPopulation/mqtt_iotprovisioning_packetPopulate.c:
```diff
@@ -56,76 +57,103 @@ void dps_client_register(uint8_t* topic, uint8_t* payload)
+    }
 
-    if (MQTT_CreatePublishPacket(&cloudPublishPacket) != true)
+    if (bRet == false)
```
nit:
```suggestion
    if (!bRet)
```

danewalton-msft commented on AzureIotPnpDps.X/platform/application_manager.c:
```diff
@@ -48,61 +52,68 @@ void  wifiConnectionStateChanged(uint8_t status);
-		mode = WIFI_SOFT_AP;
+		// AP Mode is not enabled
+		//SW0 only 
+		//debug_printGood("   APP: Starting WIFI_SOFT_AP");
+		//mode = WIFI_SOFT_AP;
```
is this supposed to be commented out?

daisukeiot replied:
AP mode wasn't working, but I will double check

daisukeiot replied:
AP Mode was never working but keeping them as comment in case someone would like to add/enable

danewalton-msft commented on AzureIotPnpDps.X/platform/application_manager.c:
```diff
@@ -176,37 +183,23 @@ uint32_t MAIN_dataTask(void* payload)
-			LED_GREEN_SetHigh();
-		}
-	}
+//	if (shared_networking_params.haveAPConnection)
+//	{
+		// Add code for periodic operation when AP mode is enabled and have connection 
+//	}
```
just checking that this is also supposed to be commented

daisukeiot replied:
This if section was there as a placeholder but there is no code to execute.  Keeping them with comment out as a placeholder.

danewalton-msft commented on AzureIotPnpDps.X/platform/cli/cli.c:
```diff
@@ -130,66 +156,65 @@ uint32_t CLI_task(void* param)
             {
-               command_received((char*)command);
-			   cmd_rcvd = true;
+                if( endOfLineTest(c) && !cmd_rcvd )
```
nit
```suggestion
                if(endOfLineTest(c) && !cmd_rcvd)
```

danewalton-msft commented on AzureIotPnpDps.X/platform/cli/cli.c:
```diff
@@ -309,11 +341,186 @@ static void get_cli_version(char *pArg)
+static void get_set_dps_idscope(char *pArg)
+{
+    char *dps_param;
+    uint8_t i;
```
nit: keep this closer to where it is used (line 505)

daisukeiot replied:
done

danewalton-msft commented on AzureIotPnpDps.X/platform/led.c:
```diff
@@ -1,228 +1,380 @@
+    {
+        case LED_BLUE:
+            LATCbits.LATC5 ^= 1; /* toggle LED_BLUE output */
+            timeout = led_status.state_flag.blue == LED_STATE_BLINK_FAST ?LED_100ms_INTERVAL:LED_400ms_INTERVAL;
```
```suggestion
            timeout = led_status.state_flag.blue == LED_STATE_BLINK_FAST ? LED_100ms_INTERVAL : LED_400ms_INTERVAL;
```

danewalton-msft commented on AzureIotPnpDps.X/platform/led.c:
```diff
@@ -1,228 +1,380 @@
+            break;
+        case LED_GREEN:
+            LATCbits.LATC4 ^= 1; /* toggle LED_GREEN output */
+            timeout = led_status.state_flag.green == LED_STATE_BLINK_FAST ?LED_100ms_INTERVAL:LED_400ms_INTERVAL;
```
```suggestion
            timeout = led_status.state_flag.green == LED_STATE_BLINK_FAST ? LED_100ms_INTERVAL : LED_400ms_INTERVAL;
```

danewalton-msft commented on AzureIotPnpDps.X/platform/led.c:
```diff
@@ -1,228 +1,380 @@
+            break;
+        case LED_YELLOW:
+            LATCbits.LATC3 ^= 1; /* toggle LED_YELLOW output */
+            timeout = led_status.state_flag.yellow == LED_STATE_BLINK_FAST ?LED_100ms_INTERVAL:LED_400ms_INTERVAL;
```
```suggestion
            timeout = led_status.state_flag.yellow == LED_STATE_BLINK_FAST ? LED_100ms_INTERVAL : LED_400ms_INTERVAL;
```

danewalton-msft commented on AzureIotPnpDps.X/platform/led.c:
```diff
@@ -1,228 +1,380 @@
+            break;
+        case LED_RED:
+            LATBbits.LATB4 ^= 1; /* toggle LED_RED output */
+            timeout = led_status.state_flag.red == LED_STATE_BLINK_FAST ?LED_100ms_INTERVAL:LED_400ms_INTERVAL;
```
```suggestion
            timeout = led_status.state_flag.red == LED_STATE_BLINK_FAST ? LED_100ms_INTERVAL : LED_400ms_INTERVAL;
```

danewalton-msft reviewed (changes requested):
There are a few places where new PnP APIs can be used to simplify the user code. I believe they were in the code before this change so requesting that they be used again here to show customers how to use them. Otherwise looks pretty good to me thanks for doing all this work.

Edit: Going to test this on my board tomorrow.

danewalton-msft commented:
I just tried running it and I keep getting an error saying that the ID scope is wrong. The instructions [here](https://github.com/daisukeiot/Microchip-PIC-IoT-Wx-1/blob/main/README.md#section-4-connect-the-pic-iot-device-to-azure) still show that the `PROVISIONING_ID_SCOPE` should be used in the `IoT_Sensor_Node_config.h`. But if I do a "find all", `PROVISIONING_ID_SCOPE` doesn't seem to be used anymore anywhere.

ericwolz commented:
PROVISIONING_ID_SCOPE is not used anymore. The scope is now stored in the NVRAM. You need to set it via the CLI. I think we should have a #define that overrides this and customer may want the DPS as part of the firmware.

danewalton-msft commented:
> 
> 
> PROVISIONING_ID_SCOPE is not used anymore. The scope is now stored in the NVRAM. You need to set it via the CLI. I think we should have a #define that overrides this and customer may want the DPS as part of the firmware.

Yup sounds good to me.

Commit:
pick up from main

daisukeiot committed:
Update AzureIotPnpDps.X/main.c

Co-authored-by: Dane Walton <dane.walton@microsoft.com>

daisukeiot committed:
Update AzureIotPnpDps.X/main.c

Co-authored-by: Dane Walton <dane.walton@microsoft.com>

daisukeiot committed:
Update AzureIotPnpDps.X/main.c

Co-authored-by: Dane Walton <dane.walton@microsoft.com>

daisukeiot committed:
Update AzureIotPnpDps.X/platform/led.c

Co-authored-by: Dane Walton <dane.walton@microsoft.com>

daisukeiot committed:
Update AzureIotPnpDps.X/platform/led.c

Co-authored-by: Dane Walton <dane.walton@microsoft.com>

daisukeiot committed:
Update AzureIotPnpDps.X/platform/led.c

Co-authored-by: Dane Walton <dane.walton@microsoft.com>

daisukeiot committed:
Update AzureIotPnpDps.X/platform/led.c

Co-authored-by: Dane Walton <dane.walton@microsoft.com>

daisukeiot committed:
Update AzureIotPnpDps.X/main.c

Co-authored-by: Dane Walton <dane.walton@microsoft.com>

daisukeiot committed:
Update AzureIotPnpDps.X/main.c

Co-authored-by: Dane Walton <dane.walton@microsoft.com>

ericwolz commented:
@daisukeiot Hey, just checking in. Is there an ETA on the changes that @danewalton requested?

ryanwinter commented:
What's the ETA on this PR? Is there a plan to reflect this into the SAM samples too?

daisukeiot committed:
Update AzureIotPnpDps.X/mqtt_packetPopulation/mqtt_iotprovisioning_packetPopulate.c

Co-authored-by: Dane Walton <dane.walton@microsoft.com>

daisukeiot committed:
Update main.c

daisukeiot committed:
use APIs

az_iot_pnp_client_property_builder_begin_reported_status() and az_iot_pnp_client_property_builder_end_reported_status()

daisukeiot committed:
Update AzureIotPnpDps.X/main.c

Co-authored-by: Dane Walton <dane.walton@microsoft.com>

daisukeiot committed:
From Review

- Rename variables with _span suffix
- az_result with 0x%08x / az_ret

daisukeiot committed:
Merge branch 'main' of https://github.com/daisukeiot/Microchip-PIC-IoT-Wx-1 into main

daisukeiot committed:
switch to az_iot_pnp_client_property_get_next_component_property()

daisukeiot committed:
Update AzureIotPnpDps.X/main.c

Co-authored-by: Dane Walton <dane.walton@microsoft.com>

daisukeiot committed:
Update AzureIotPnpDps.X/main.c

Co-authored-by: Dane Walton <dane.walton@microsoft.com>

daisukeiot committed:
make booleans with verbs

daisukeiot committed:
consistent code style

daisukeiot committed:
Update AzureIotPnpDps.X/platform/cli/cli.c

Co-authored-by: Dane Walton <dane.walton@microsoft.com>

daisukeiot commented:
I had to reduce code size to fit.  Almost there..

daisukeiot committed:
Update cli.c

daisukeiot committed:
Merge branch 'main' of https://github.com/daisukeiot/Microchip-PIC-IoT-Wx-1 into main

daisukeiot committed:
Update cloud_service.c

daisukeiot committed:
Revert "Update cloud_service.c"

This reverts commit b1fdc3b2e0cbd2a16684e239efd42cf09ab33abc.

daisukeiot committed:
Revert "pick up from main"

This reverts commit 9effd1b95405d3f18bb73737514d53adcfb4fcd5.

daisukeiot committed:
Fix indent

daisukeiot committed:
Picking up change and disabling LED CLI command

danewalton-msft reviewed (approved):
Unblocking since the previous pnp api shortcomings are fixed. I will try it on my board to double check but the content looks good. Thanks again @daisukeiot for all the work. This was a great amount of work.

danewalton-msft commented:
@daisukeiot have you tried compiling the project with `SAS` enabled? I think the build in that case is broken.

danewalton-msft commented:
Can confirm that it works. The README should be updated to include the new method of inputting the ID Scope through the CLI and that the AP name should be updated not the SSID.

daisukeiot committed:
Merge branch 'main' into main

daisukeiot committed:
Doc updates

daisukeiot committed:
Update README.md

daisukeiot committed:
Disable LED CLI

daisukeiot commented:
@danewalton
Per Microchip, there is a separate repo for SAS token so it's ok for this repo to be X.509 only.
